sdk/core/azure-core/src/main/java/com/azure/core/util/RequestContent.java

```diff
@@ -0,0 +1,248 @@
+     *
+     * @param requestOutbound The outbound where the request will be written.
+     * @return An asynchronous response which will emit once the request has completed writing.
+     */
+    default Mono<Void> writeToAsync(RequestOutbound requestOutbound) {
+        return Mono.defer(() -> Mono.fromRunnable(() -> writeTo(requestOutbound)));
+    }
+
+    /**
+     * Write the {@link RequestContent} to the {@link RequestOutbound}.
+     *
+     * @param requestOutbound The outbound where the request will be written.
+     */
+    void writeTo(RequestOutbound requestOutbound);
+
+    /**
+     * Converts the {@link RequestContent} into a {@code Flux<ByteBuffer>} for use in reactive streams.
+     *
+     * @return The {@link RequestContent} as a {@code Flux<ByteBuffer>}.
+     */
+    Flux<ByteBuffer> asFluxByteBuffer();
+
+    /**
+     * Gets the length of the {@link RequestContent} if it is able to be calculated.
+     * <p>
+     * If the content length isn't able to be calculated null will be returned.
+     *
+     * @return The length of the {@link RequestContent} if it is able to be calculated, otherwise null.
+     */
+    Long getLength();
+
+    /**
+     * Creates a {@link RequestContent} that uses {@code byte[]} as its data.
+     *
+     * @param bytes The bytes that will be the {@link RequestContent} data.
+     * @return A new {@link RequestContent}.
+     * @throws NullPointerException If {@code bytes} is null.
+     */
+    static RequestContent create(byte[] bytes) {
+        Objects.requireNonNull(bytes, "'bytes' cannot be null.");
+        return create(bytes, 0, bytes.length);
+    }
+
+    /**
+     * Creates a {@link RequestContent} that uses {@code byte[]} as its data.
+     *
+     * @param bytes The bytes that will be the {@link RequestContent} data.
+     * @param offset Offset in the bytes where the data will begin.
+     * @param length Length of the data.
+     * @return A new {@link RequestContent}.
+     * @throws NullPointerException If {@code bytes} is null.
+     */
+    static RequestContent create(byte[] bytes, int offset, int length) {
+        Objects.requireNonNull(bytes, "'bytes' cannot be null.");
+        return new ArrayContent(bytes, offset, length);
+    }
+
+    /**
+     * Creates a {@link RequestContent} that uses {@link String} as its data.
+     * <p>
+     * The passed {@link String} is converted using {@link StandardCharsets#UTF_8}, if another character set is required
+     * use {@link #create(byte[])} and pass {@link String#getBytes(Charset)} using the required character set.
+     *
+     * @param content The string that will be the {@link RequestContent} data.
+     * @return A new {@link RequestContent}.
+     * @throws NullPointerException If {@code content} is null.
+     */
+    static RequestContent create(String content) {
+        Objects.requireNonNull(content, "'content' cannot be null.");
+        return create(content.getBytes(StandardCharsets.UTF_8));
```

Comment:
Does the `byte[]` need to be created eagerly? Instead could there be another `RequestContent` implementation which defers this until `writeTo` time.